libs/core/src/lib/token/tokenizer.component.ts

```diff
@@ -482,4 +515,82 @@ export class TokenizerComponent implements AfterViewChecked, AfterViewInit, Afte
             });
         }
     }
+
+    private inputKeydownEvent(): void {
+      this.input.elementRef().nativeElement.addEventListener('keydown', (event) => {
+        this.handleKeyDown(event, this.tokenList.length);
+      });
+    }
+
+    private tokenizerKeyupEvent(): void {
+      this.tokenizerInnerEl.nativeElement.addEventListener('keyup', (event) => {
+        this.handleKeyUp(event);
+      });
+    }
+
+    private basicSelected(token, index): void {
+      this.tokenList.forEach(shadowedToken => {
+        if (shadowedToken !== token) {
+            shadowedToken.selected = false;
+        }
+    });
+    this.firstElement = index;
+    this.lastElement = index;
+    token.selected = true;
+    }
+
+    private shiftSelected(index): void {
+
+      if (index < this.firstElement) {
+        this.directionRight = false;
+        this.firstElement = index;
+      } else if (index > this.lastElement) {
+        this.directionRight = true;
+        this.lastElement = index;
+      }
+      if (this.lastElement === this.firstElement) {
+        this.directionRight = null;
+      }
+      if (!this.directionRight) {
+        this.firstElement = index;
+      } else if (this.directionRight) {
+        this.lastElement = index;
+      }
+
+      this.tokenList.forEach((token, indexOfToken) => {
+        if (indexOfToken >= this.firstElement && indexOfToken <= this.lastElement) {
+          token.selected = true;
+        } else {
+          token.selected = false;
+        }
+      });
+    }
+
+    private ctrlSelected(token, index): void {
```

Comment:
Let's add some comments into the code, I don't understand what is going on in this method